compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -2791,11 +2791,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     )
                 })
                 .map(|mut field_path| {
+                    use std::fmt::Write;
+
                     field_path.pop();
-                    field_path
-                        .iter()
-                        .map(|id| format!("{}.", id.name.to_ident_string()))
-                        .collect::<String>()
+                    field_path.iter().fold(String::new(), |mut s, id| {
+                        write!(s, "{}.", id.name.to_ident_string()).unwrap();
+                        s
+                    })
```

Comment:
This feels objectively worse... :/

Reply:
This is also in diagnostics code, so maybe we just want to `allow` it...

Reply:
Yeah, it doesn't seem great for diagnostics code. To be honest, unless we decide to gate CI on the perf lints, I would perhaps just rather close this PR, since these three fixes don't bring any benefits, and unless CI checks it, new lint warnings might appear anytime in the future...

Reply:
Yeah, ok, I guess let's close it for now. Happy to review a PR that unbreaks CI in case we do end up needing to apply these changes.